Add draft nycOpenData blog post (#1341)
ropensci/roweb3: content/blog/2026-08-06-the-journey-of-nycopendata-from-classroom-to-community/index.md

---
title: "The Journey of {nycOpenData}: From Classroom to Community"
author: Christian Martinez
date: "2026-08-06"
slug: the-journey-of-nycopendata-from-classroom-to-community
package_version: 0.2.3
categories: []
tags:
  - open-data
  - open-source
  - R
  - education
  - package-development
  - community
  - API
---

## FADE IN

> Nine students in the M.S. in Psychology Research program at Brooklyn College are taking the course *Reproducible Psychological Research*, where they are learning how to code for the first time. All of them have their cameras off. They are all learning R for the first time, and as an introduction to R, are being taught using the `mtcars` dataset.

Reviewer: I just remembered that we should add some context for those less familiar with the US.
```suggestion
Nine students in the M.S. in Psychology Research program at Brooklyn College, New York City (NYC), USA are taking the course *Reproducible Psychological Research*, where they are learning how to code for the first time. All of them have their cameras off. They are all learning R for the first time, and as an introduction to R, are being taught using the `mtcars` dataset.
```

This not only made teaching with NYC Open Data easier, but more fun too. The package did exactly what it needed to do: make working with NYC Open Data datasets incredibly easy in R.

With this version of `nycOpenData`, students were able to complete their final projects with one simple prompt:

> Answer any question you have about NYC using open data.

Those projects eventually became the [NYC Open Data Student Gallery Book](https://martinezc1-nyc-open-data-student-gallery.share.connect.posit.cloud/), which was presented at [NYC Open Data Week 2026](https://www.youtube.com/watch?v=Vw8PhY7NDGE). Each student's project became its own chapter, and each student had the opportunity to present their work to the world. These same students who were once falling asleep while I basically sang them a lullaby, now were very much awake, displaying what they've learned and uncovered to an audience of professionals.

As an educator, my goal is to help my students be as successful as possible. Through `nycOpenData`, that goal became a reality.

All of the work I put into this package was worth it.

> ## `nycOpenData`: Student Contributions

Reviewer: ```suggestion
## `nycOpenData`: Student contributions
```

The following semester, I had the opportunity to teach a follow-up class to the same cohort. Then came another crazy idea:

> What if the same students who used `nycOpenData` to not only learn, but also conduct research, contributed to the same infrastructure they utilized?

Basically, what if I helped my students contribute to the `nycOpenData` package?

This would be similar to what a junior software developer might do. Using my code as a template, they could build their own functions for whatever dataset they were interested in. The package already had 30 functions---what's another nine?

Through light introductions to GitHub, package development, and open-source software, each of my students successfully contributed their own function to the `nycOpenData` package. Suddenly, they weren't just users of open-source software---they were contributors.

Good thing I didn't do this last semester, as that would have guaranteed brain explosions from my students.

## `nycOpenData`: rOpenSci

> As time progressed, and as I assessed the student contributions, I became more familiar with package development. Wanting the best product possible, I decided to submit the package to rOpenSci through its [**public peer review process**](https://github.com/ropensci/software-review/issues/750#issuecomment-5007655734). I was a first-time package developer and wanted guidance from people who had far more experience than I did. I wanted anyone---whether my students or someone using the package halfway around the world---to have the best experience possible. After all, the goal was to make NYC Open Data more accessible for *anyone*. To go from helping just my students, to my students helping, to potentially helping the overall R community through rOpenSci---I would have never guessed this trajectory.

Reviewer: ```suggestion
As time progressed, and as I assessed the student contributions, I became more familiar with package development. Wanting the best product possible, I decided to submit the package to rOpenSci through its [**public peer review process**](https://github.com/ropensci/software-review/issues/750#issuecomment-5007655734). I was a first-time package developer and wanted guidance from people who had far more experience than I did. I wanted anyone---whether my students or someone using the package halfway around the world---to have the best experience possible. After all, the goal was to make NYC Open Data more accessible for *anyone*. To go from helping just my students, to my students just helping me, to potentially helping the overall R community through rOpenSci---I would have never guessed this trajectory.
```

This ended up being one of the best experiences I've had as a programmer.

The review process didn't simply determine whether the package was ready---it made the package better. Every suggestion strengthened the documentation, improved the user experience, and taught me something new about developing open-source software.

**First of all, my editor and reviewers told me exactly what I needed to hear:**

> Having 39 functions that basically do the same thing is not only unmanageable, but bad code.

Which was 100% the truth.

And, if I'm being honest, the package itself came about while I was teaching a course on reproducibility. I was literally teaching my students that copy-and-paste was *not* something to do, while I was doing the exact same thing. Sure, I brushed it off because I needed to put something together quickly, but it was time to face the facts and fix it.

With that came good news and bad news. As for every action, there is an equal and opposite reaction:

- The *good* news: I was able to turn the 39 functions into just 3. Now, all 2,000+ datasets can be pulled from the portal. Easier to maintain, easier for users to understand, and overall a much better product.

> - The *bad* news: All of my students' work was now gone. (Sad face.)

Reviewer: I forgot to mention this in the first round, feel free to use an emoji here if you want to, or leave it as is, if that's the style you prefer (I wasn't sure if it was a style choice or not). 

```suggestion
- The *bad* news: All of my students' work was now gone. 🙁
```

Thanks to the thoughtful feedback and hard work of the editor, reviewers, and everyone involved in the review process, `nycOpenData` was successfully accepted into rOpenSci---an indescribable privilege. I honestly can't believe the journey from where the package started to where it is today.

> I am especially grateful to my editor, [**\@ronnyhdez**](https://github.com/ronnyhdez), and reviewers, [**\@donghl17**](https://github.com/donghl17) and [**\@MichaelPascale**](https://github.com/MichaelPascale), for the time, care, and thoughtful feedback they invested throughout the review. As a first-time package developer, I couldn't have asked for a better introduction to open-source software development.

Reviewer: Nice to use names if they're available, and link to rOpenSci author pages if they have them.
```suggestion
I am especially grateful to my editor, [**Ronny Hernandez Mora**](https://ropensci.org/author/ronny-hernandez-mora/), and reviewers, [**\@donghl17**](https://github.com/donghl17) and [**Michael Pascale**](https://github.com/MichaelPascale), for the time, care, and thoughtful feedback they invested throughout the review. As a first-time package developer, I couldn't have asked for a better introduction to open-source software development.
```

With the foundation finally built, another thought entered my mind:

> Can we use this same infrastructure for other open data portals?

Turns out, you can!

## The OpenData Ecosystem

With the infrastructure in place, another realization hit me: I never had to build it from scratch again.

The same foundation that powered `nycOpenData` could be adapted to other open data portals, making it possible to create an entire ecosystem of OpenData packages.

> If students in NYC were motivated by using data regarding their own city, why wouldn't the same hold true for students in other cities?

Since then, four additional packages have been built, turning one package into an ecosystem:

- `nysOpenData` --- New York State Open Data
- `mtaOpenData` --- Metropolitan Transportation Authority Open Data
- `laOpenData` --- Los Angeles Open Data
- `chiOpenData` --- Chicago Open Data

> - `ausOpenData` --- Austin Open Data

Reviewer: I missed this, it'd be great to add links to the different repositories or documentation sites so people can easily find them.

Author: Done!

> The ecosystem continues to grow. A former student and I recently collaborated on the `ausOpenData` package (which was just recently accepted to CRAN), and a current intern at the [NYC Open Data Lab](https://nycopendatalab.org) is building `sfOpenData`. This not only gives others an opportunity to contribute, but also gives me a chance to continue involving students in the ecosystem after the original `nycOpenData` functions were "refactored."

Reviewer: Also add link for sfOpenData!

Hopefully, there are many more to come. The hard part is no longer building packages---it is deciding which open data portal comes next.

## `nycOpenData`: The Final Product

What started as a single function has turned into six OpenData packages, one of which has been successfully peer-reviewed by rOpenSci, with the others built on the same foundation. *Tens of billions* of rows of public data, from across the country, are now easily accessible in R for anyone---from researchers and journalists to students and educators.

Looking back, I think this journey embodies the spirit of R.

One of the best things about R is that it is open source. Anyone with an idea can contribute, and those ideas have a way of growing into something much bigger than originally imagined. This entire ecosystem started because nine graduate students didn't care about the `mtcars` dataset.

> On top of that, the tools used to build this ecosystem were free and open source, and the packages themselves are freely available to anyone. In my family we say, *"If it's for free, it's for me."*

Reviewer: Great rewording 😸
